crates/bevy_asset/src/io/android.rs

```diff
@@ -56,18 +55,47 @@ impl AssetReader for AndroidAssetReader {
 
     fn read_directory<'a>(
         &'a self,
-        _path: &'a Path,
+        path: &'a Path,
     ) -> BoxedFuture<'a, Result<Box<PathStream>, AssetReaderError>> {
-        let stream: Box<PathStream> = Box::new(EmptyPathStream);
-        error!("Reading directories is not supported with the AndroidAssetReader");
-        Box::pin(async move { Ok(stream) })
+        Box::pin(async move {
+            let asset_manager = bevy_winit::ANDROID_APP
+                .get()
+                .expect("Bevy must be setup with the #[bevy_main] macro on Android")
+                .asset_manager();
+            let opened_assets_dir = asset_manager
+                .open_dir(&CString::new(path.to_str().unwrap()).unwrap())
+                .ok_or(AssetReaderError::NotFound(path.to_path_buf()))?;
+
+            let mapped_stream = opened_assets_dir
+                .filter_map(move |f| {
+                    let file_path = path.join(Path::new(f.to_str().unwrap()));
+                    // filter out meta files as they are not considered assets
+                    if let Some(ext) = file_path.extension().and_then(|e| e.to_str()) {
+                        if ext.eq_ignore_ascii_case("meta") {
+                            return None;
+                        }
+                    }
+                    Some(file_path.to_owned())
+                })
+                .collect::<Vec<_>>();
+
+            let read_dir: Box<PathStream> = Box::new(stream::iter(mapped_stream));
+            Ok(read_dir)
+        })
     }
 
     fn is_directory<'a>(
         &'a self,
-        _path: &'a Path,
+        path: &'a Path,
     ) -> BoxedFuture<'a, std::result::Result<bool, AssetReaderError>> {
-        error!("Reading directories is not supported with the AndroidAssetReader");
-        Box::pin(async move { Ok(false) })
+        Box::pin(async move {
+            let asset_manager = bevy_winit::ANDROID_APP
+                .get()
+                .expect("Bevy must be setup with the #[bevy_main] macro on Android")
+                .asset_manager();
+            Ok(asset_manager
+                .open(&CString::new(path.to_str().unwrap()).unwrap())
+                .map_or(true, |_| false))
```

Comment:
Yeah, unfortunately it will. I haven't found a non hacky way to implement `is_directory`. There is no way to do that check using the asset manager and using `open_dir` was successful even for non directories, but `open` failed for directories, so that was pretty much the only thing I had available. I assumed that it was mostly used right before using `read_directory`, so I thought it was fine to just not return these errors.

We could do it a bit differently by first making a call to `open_dir` and propagating whatever error we get from that, before making this call. That would cover the cases of non existing file/directory. Would that make sense?

Reply:
Sure, that seems more robust that way 🙂

Seems like there isn't any error exposed by the ndk for the asset related functions, maybe there would be a way to get a proper error by calling into the java `AssetManager` instance instead of the ndk one? Anyway what you suggested seems to fit to avoid getting false positives on a directory existence.